src/AggregateFunctions/AggregateFunctionMap.h

```diff
@@ -0,0 +1,207 @@
+struct AggregateFunctionMapCombinatorData
+{
+    std::unordered_map<KeyType, AggregateDataPtr> merged_maps;
+
+    static void writeKey(KeyType key, WriteBuffer & buf) { writeBinary(key, buf); }
+    static void readKey(KeyType & key, ReadBuffer & buf) { readBinary(key, buf); }
+};
+
+template <>
+struct AggregateFunctionMapCombinatorData<String>
+{
+    std::unordered_map<String, AggregateDataPtr> merged_maps;
+
+    static void writeKey(String key, WriteBuffer & buf) { writeString(key, buf); }
+    static void readKey(String & key, ReadBuffer & buf) { readString(key, buf); }
+};
+
+template <typename KeyType>
+class AggregateFunctionMap final
+    : public IAggregateFunctionDataHelper<AggregateFunctionMapCombinatorData<KeyType>, AggregateFunctionMap<KeyType>>
+{
+private:
+    DataTypePtr key_type;
+    AggregateFunctionPtr nested_func;
+    using Base = IAggregateFunctionDataHelper<AggregateFunctionMapCombinatorData<KeyType>, AggregateFunctionMap<KeyType>>;
+
+public:
+    AggregateFunctionMap(AggregateFunctionPtr nested, const DataTypes & types) : Base(types, nested->getParameters()), nested_func(nested)
+    {
+        if (types.empty())
+            throw Exception(
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Aggregate function " + getName() + " requires at least one argument");
+
+        if (types.size() > 1)
+            throw Exception(
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Aggregate function " + getName() + " requires only one map argument");
+
+        const auto * map_type = checkAndGetDataType<DataTypeMap>(types[0].get());
+        if (!map_type)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Aggregate function " + getName() + " requires map as argument");
+
+        key_type = map_type->getKeyType();
+    }
+
+    String getName() const override { return nested_func->getName() + "Map"; }
+
+    DataTypePtr getReturnType() const override { return std::make_shared<DataTypeMap>(DataTypes{key_type, nested_func->getReturnType()}); }
+
+    void add(AggregateDataPtr place, const IColumn ** columns, size_t row_num, Arena * arena) const override
+    {
+        const auto & map_column = assert_cast<const ColumnMap &>(*columns[0]);
+        const auto & map_nested_tuple = map_column.getNestedData();
+        const IColumn::Offsets & map_array_offsets = map_column.getNestedColumn().getOffsets();
+
+        const size_t offset = map_array_offsets[row_num - 1];
+        const size_t size = (map_array_offsets[row_num] - offset);
+
+        const auto & key_column = map_nested_tuple.getColumn(0);
+        const auto & val_column = map_nested_tuple.getColumn(1);
+
+        auto & merged_maps = this->data(place).merged_maps;
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            KeyType key;
+            if constexpr (std::is_same<KeyType, String>::value)
+            {
+                key = key_column.operator[](offset + i).get<KeyType>();
```

Comment:
It seems like we make double copy of a string here. One on creating Field:
https://github.com/ClickHouse/ClickHouse/blob/38493bde487ce48e37f0fdbcb2894f55d78298c3/src/Core/Field.h#L677
And the second when we perform `operator =` for String. Can we just cast `key_column` to `ColumnString` and use `getData` method and use returned StringRef to construct String?

Reply:
I have some problems with const columns if I use direct casting, investigating

Reply:
Should be fixed, in current version String construction is avoided as much as possible.